prov/opx/include/rdma/opx/fi_opx_hmem.h

```diff
@@ -107,6 +120,74 @@ enum fi_hmem_iface fi_opx_hmem_get_iface(const void *ptr,
 	return FI_HMEM_SYSTEM;
 }
 
+__OPX_FORCE_INLINE__
+int opx_copy_to_hmem(enum fi_hmem_iface iface, uint64_t device, uint64_t hmem_handle,
+		     void *dest, const void *src, size_t len, int64_t threshold)
+{
+	// These functions should never be called for regular host memory.
+	// Calling this function directly should only ever be done in code
+	// paths where we know iface != FI_HMEM_SYSTEM. Otherwise, the
+	// OPX_HMEM_COPY_* macros should be used
+	assert(iface != FI_HMEM_SYSTEM);
+
+	int ret;
+	assert((hmem_handle == OPX_HMEM_NO_HANDLE && threshold == OPX_HMEM_DEV_REG_THRESHOLD_NOT_SET) ||
+		(hmem_handle != OPX_HMEM_NO_HANDLE && threshold != OPX_HMEM_DEV_REG_THRESHOLD_NOT_SET));
+	OPX_TRACER_TRACE(OPX_TRACER_BEGIN, "COPY-TO-HMEM");
+#if HAVE_CUDA
+	if (hmem_handle != 0 && len <= threshold) {
+		OPX_TRACER_TRACE(OPX_TRACER_BEGIN, "GDRCOPY-TO-DEV");
+		cuda_gdrcopy_to_dev(hmem_handle, dest, src, len);
+		OPX_TRACER_TRACE(OPX_TRACER_END_SUCCESS, "GDRCOPY-TO-DEV");
+		ret = 0;
+	} else {
+		OPX_TRACER_TRACE(OPX_TRACER_BEGIN, "CUDAMEMCPY-TO-HMEM");
+		ret = (int) cudaMemcpy(dest, src, len, cudaMemcpyHostToDevice);
+		OPX_TRACER_TRACE(OPX_TRACER_END_SUCCESS, "CUDAMEMCPY-TO-HMEM");
+	}
```

Comment:
Does that mean opx only support CUDA when CUDA is enabled?

Reply:
Not exactly sure what you're asking since it's a little unclear. If you're asking, will OPX only support CUDA if it's enabled through HAVE_CUDA, the answer is, we believe so. There might be a situation in which this falls through to `ofi_copy_from_hmem` and CUDA is still used, but from what we've seen, the CUDA calls are wrapped in `#if HAVE_CUDA` in `hmem_cuda.c` as well...

If instead you're asking, can we only support CUDA at present when CUDA is enabled (and not also alternatives like ROCR at the same time), then in this PR, that's correct; however, we do have an upcoming PR that changes this to be slightly more general with an addition of `#if HAVE_ROCR` logic in addition to the `#if HAVE_CUDA` case. As we add support for additional GPUs, we can add more of that type of logic.

I intend to open that other PR before the next release, btw.

Reply:
@charlesshereda Your second paragraph answered my question. I was asking because the code block just assumes the memory is CUDA allocated.